include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -108,23 +118,216 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
+                {
+                   __lacc[__idx] = __unary_op(__in[__idx]);
+                }
+
+                __group_scan<_ValueType>(__group, __lacc.get_pointer(), __lacc.get_pointer() + __n, __bin_op, __init);
+
+                for (::uint16_t __idx = __item_id; __idx < __n; __idx += __wg_size)
+                {
+                   __out[__idx] = __lacc[__idx];
+                }
+
+                const ::std::uint16_t __residual = __n % __wg_size;
+                const ::std::uint16_t __residual_start = __n - __residual;
+                if (__residual > 0 && __item_id < __residual)
+                {
+                    auto __idx = __residual_start + __item_id;
+                   __out[__idx] = __lacc[__idx];
+                }
+            });
+        });
+        __event.wait();
+    }
+
+    template<::uint16_t _ElemsPerItem, ::uint16_t _WGSize, bool _IsFullGroup, typename _Policy, typename _InRng, typename _OutRng, typename _InitType, typename _BinaryOperation, typename _UnaryOp>
+    static void __launch_static_bounds_scan(_Policy const & __policy, _InRng&& __in, _OutRng&& __out, std::size_t __n, _InitType __init, const _BinaryOperation& __bin_op, const _UnaryOp& __unary_op)
+    {
+        using _ValueType = typename _InitType::__value_type;
+        using _CustomName = typename _Policy::kernel_name;
+        using _GroupScanKernel =
+                     oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__par_backend_hetero::__scan_single_wg_kernel, _CustomName, _BinaryOperation, std::integral_constant<bool, _Inclusive>, std::integral_constant<bool, _IsFullGroup>, _InRng, _OutRng, std::integral_constant<::std::size_t, _ElemsPerItem>, std::integral_constant<std::size_t, _WGSize>>;
+
+        constexpr ::uint32_t __elems_per_wg = _ElemsPerItem * _WGSize;
+
+        auto __event = __policy.queue().submit([&](sycl::handler& __hdl) {
+            oneapi::dpl::__ranges::__require_access(__hdl, __in, __out);
+
+            auto __lacc = __dpl_sycl::__local_accessor<_ValueType>(sycl::range<1>{__elems_per_wg}, __hdl);
+
+            __hdl.parallel_for<__par_backend_hetero::__i_kernel_name<_GroupScanKernel, __elems_per_wg, _IsFullGroup, _Inclusive>>(sycl::nd_range<1>(_WGSize, _WGSize), [=](sycl::nd_item<1> __self_item) {
+                const auto& __group = __self_item.get_group();
+                const auto& __subgroup = __self_item.get_sub_group();
+                // This kernel is only launched for sizes less than 2^16
+                const ::uint16_t __item_id = __self_item.get_local_linear_id();
+                ::uint16_t __subgroup_id = __subgroup.get_group_id();
+                ::uint16_t __id_in_subgroup = __subgroup.get_local_id();
+                ::uint16_t __subgroup_size = __subgroup.get_local_linear_range();
+
+                if constexpr (_IsFullGroup)
+                {
+                    #pragma unroll
```

Comment:
Does it make sense to check if this pragma exist? I don't exactly remember what's going on when pragma doesn't exist. I think a compiler just ignores it. I just want to make sure we don't introduce some compiler warnings/errors for the user occasionally.

Reply:
Good point. This might introduce unknown pragma warnings for some compilers. GCC, Clang and icpx understand this pragma so I can enable it just for those compilers.